### ISSUE TITLE
PEM_read_PrivateKeyfrom() doesn't work with FILE larger than 4GiB

### ISSUE THREAD

**Issue by tomato42** [openssl/openssl#23653]
I have a large file with concatenated PEM private keys, when I use the `PEM_read_PrivateKeyfrom()` function to read them one by one, after the read pointer reaches 4GiB mark, the file is rewound to the beginning, as reported by `ftell()`
(the first value is the number of read private keys, the second values is position in the file):
```
75550000: 4100256 kiB
75560000: 4116899 kiB
75570000: 4133541 kiB
75580000: 4150184 kiB
75590000: 4166827 kiB
75600000: 4183469 kiB
75610000: 5808 kiB
75620000: 22451 kiB
75630000: 39094 kiB
75640000: 55736 kiB
75650000: 72379 kiB
75660000: 89021 kiB
75670000: 105664 kiB
75680000: 122306 kiB
75690000: 138949 kiB
75700000: 155592 kiB
```

this happens with a simple loop:
```
do {
   if ((!feof(fp) &&
       (pkey = PEM_read_PrivateKey(fp, NULL, NULL, NULL) == NULL))
} while (pkey);
````
even with `#define _FILE_OFFSET_BITS 64` on a 64bit platform.

Tested with `openssl-3.0.9-2.fc38.x86_64`

**Comment by tomato42** [comment]
it's not about reading large private keys, it's about reading private keys from inside a file that's large

but yeah, given that nobody else hit this issue, it does sound like it's more of an obscure problem

**Comment by t8m** [comment]
Yeah, I understand that this is not about a huge private key alone in the file but a regular private key at some large file offset. I assume not just private keys but other data decoded by decoders would be affected (params, public keys). I think certificates or CRLs should not be affected but I did not test this.

Still I consider this to be an extremely unlikely use case.

**Comment by nhorman** [comment]
FWIW, looks like it might be an issue in pem_read_bio_key(). It calls BIO_tell and BIO_seek using an int, which will certainly overflow at 4GB, even on a 64bit platform.  I'd try changing the type of pos in that function to long and see if that doesn't fix it.  Note the same change likely needs to be made in pem_read_bio_key_decoder()

**Comment by kroeckx** [comment]
Note that long is also only 32 bit on most platforms. If you change it, it should probably be an  off_t, and we might need to change calls to functions like fseeko and ftello.

**Comment by tomato42** [comment]
@kroeckx that's not what the description of `_LARGEFILE64_SOURCE`, `_LARGEFILE_SOURCE`, and `_FILE_OFFSET_BITS` in `feature_test_macros(7)` says. Using `long` on 64 bit platform should be sufficient.

**Comment by kroeckx** [comment]
Long on windows is still 32 bit.

**Comment by kroeckx** [comment]
Reading that manpage seems to say exactly the same, use off_t, and the new functions.

**Comment by davidben** [comment]
I think you just do:

1. `BIO_C_FILE_TELL_EX` and `BIO_C_FILE_SEEK_EX` return/take the `uint64_t` offset via the `void*`. (Pass the address of the `uint64_t`.
2. `BIO_tell_ex` and `BIO_seek_ex` are functions, *not* macros. (See https://github.com/openssl/openssl/issues/1319, filed in 2016, and originally on the radar for post-1.1.0.) They first try the new ctrls, and if those fail, call the new ones.
3. Replace `BIO_tell` and `BIO_seek` with wrappers over `BIO_tell_ex` and `BIO_seek_ex`

Not too bad, but do remember to make sure the full product of {old caller, new caller} x {old BIO, new BIO} works. OpenSSL neglected to do this exercise for `BIO_read_ex` and the result is the giant mess in https://github.com/openssl/openssl/issues/8208. To this day, EOF-handling in `BIO`s is ambiguous, undocumented, and a compatibility break. (Instead of adding `BIO_eof` to built-in `BIO`s one-by-one, `BIO_ctrl` should have simulated `BIO_eof` based on `BIO_read`'s return value. That's the only way to make the new convention work with existing applications.)

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
